atom/browser/atom_browser_main_parts.cc

```diff
@@ -200,6 +191,32 @@ void AtomBrowserMainParts::PostMainMessageLoopStart() {
 #endif
   device::GeolocationProvider::SetGeolocationDelegate(
       new AtomGeolocationDelegate());
+
+  base::FilePath user_dir;
+  PathService::Get(brightray::DIR_USER_DATA, &user_dir);
+  process_singleton_.reset(new ProcessSingleton(user_dir));
+
+  js_env_.reset(new JavascriptEnvironment);
```

Comment:
We can not delay the initialization of Node.js to `PostMainMessageLoopStart`, it is too late for certain APIs to work. (Like changing default locations for Chromium dirs.)

Reply:
I'll backport https://github.com/electron/electron/pull/11230/commits/ad493c2654d9eb7b5fe5c779b8fc4f8a3b29d9a8 to the master branch to fix the V8 dead lock on startup, which should be able to fix some dcheck failures here.

Reply:
> We can not delay the initialization of Node.js to PostMainMessageLoopStart, it is too late for certain APIs to work. (Like changing default locations for Chromium dirs.)

The DCHECK here https://github.com/electron/electron/pull/11396/files#diff-85bd8d8e80263b8547de573fd87cb532R532 fails with being called on invalid thread instead of main thread otherwise. Any ideas on how to fix it ?

Reply:
That's a surprise to me, it is impossible for `App` to be initialized on threads other than UI thread, is it because the `BrowserThread` class has not been initialized by the time the Node.js modules are loaded?

Regarding the task runner, by the time of `JavaScriptEnvironment`'s initialization, there is no thread task runner available, I used `BridgeTaskRunner` to workaround it.

Reply:
Yup, the main thread is initialized only in [BrowserMainLoop::MainMessageLoopStart](https://cs.chromium.org/chromium/src/content/browser/browser_main_runner.cc?sq=package:chromium&l=107) , while the node modules are loaded early on because of `PostEarlyInitialization` hook. I am not sure of the best way to fix this and at the same time not cause any performance issue.

Reply:
I think we can just remove the `DCHECK_CURRENTLY_ON` in `App:App`.

Reply:
Cool but is it ok for the app module to be initialized on a different thread that UI ?

Reply:
I don't think the module is initialized on other threads, the DCHECK failed because the `BrowserThread` class is not initialized, not because `App` is initialized on other threads.

Reply:
👍

Reply:
Even if I remove the DCHECK on app module, I hit the DCHECK in [BrowserChildProcessObserver::Add](https://github.com/electron/electron/blob/master/atom/browser/api/atom_api_app.cc#L535)

```
[36149:1222/174140.514620:FATAL:browser_child_process_host_impl.cc(133)] Check failed: ::content::BrowserThread::CurrentlyOn(BrowserThread::UI). Must be called on Chrome_UIThread; actually called on Unknown Thread.
0   libbase.dylib                       0x000000011242b60e base::debug::StackTrace::StackTrace(unsigned long) + 174
1   libbase.dylib                       0x000000011242b6dd base::debug::StackTrace::StackTrace(unsigned long) + 29
2   libbase.dylib                       0x000000011242997c base::debug::StackTrace::StackTrace() + 28
3   libbase.dylib                       0x00000001124c84af logging::LogMessage::~LogMessage() + 479
4   libbase.dylib                       0x00000001124c5e25 logging::LogMessage::~LogMessage() + 21
5   libcontent.dylib                    0x000000011b0e95ee content::BrowserChildProcessHostImpl::AddObserver(content::BrowserChildProcessObserver*) + 270
6   libcontent.dylib                    0x0000000119aae155 content::BrowserChildProcessObserver::Add(content::BrowserChildProcessObserver*) + 21
7   Electron Framework                  0x000000010e259324 atom::api::App::App(v8::Isolate*) + 916
8   Electron Framework                  0x000000010e25acbd atom::api::App::App(v8::Isolate*) + 29
9   Electron Framework                  0x000000010e263233 atom::api::App::Create(v8::Isolate*) + 51
10  Electron Framework                  0x000000010e2692d2 (anonymous namespace)::Initialize(v8::Local<v8::Object>, v8::Local<v8::Value>, v8::Local<v8::Context>, void*) + 258
11  libnode.dylib                       0x0000000110acb72b node::InitModule(node::Environment*, node::node_module*, v8::Local<v8::String>) + 251
12  libnode.dylib                       0x0000000110ac4254 node::Binding(v8::FunctionCallbackInfo<v8::Value> const&) + 692
13  libv8.dylib                         0x00000001116b6292 v8::internal::StrDup(char const*) + 722
14  libv8.dylib                         0x0000000111788615 v8::internal::operator<<(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, v8::internal::BasicBlockProfiler const&) + 137269
15  libv8.dylib                         0x0000000111787387 v8::internal::operator<<(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, v8::internal::BasicBlockProfiler const&) + 132519
16  ???                                 0x00000de827b044c4 0x0 + 15290749437124
17  ???                                 0x00000de827ca5fc0 0x0 + 15290751147968
```

should probably move it once browser thread is initialized.

Reply:
We can move the `content::BrowserChildProcessObserver::Add` call to `App::PreMainMessageLoopRun`.